Add array validator

kgrz opened the pull request:
Fixes #1

kgrz committed:
Add Array validation

To check if a given array is a member of another array, the best way
would be to convert those arrays into Sets and verifying if the given
array is a subset of the other array

kgrz committed:
Move setup classes to appropriate support files

kgrz commented:
(Forked off of the pending PR)

kgrz committed:
Fix naming conflicts

kgrz committed:
Update readme to explain usage of Array Validator

kgrz committed:
Remove rails as a dependency and add activemodel

We don't need the rest of the railties for this gem.

iffyuva commented:
okie, merge latest master onto this one again!

kgrz committed:
Merge branch 'master' into add-array-validator

iffyuva commented on README.md:
```diff
@@ -125,6 +125,56 @@ class EventForm
+    # compared to the verification array.
+
+    validates [1, 2], verify_array: { eq: [2, 1] }          # => true
+    validates [1, 2], verify_array: { eq: [3, 1] }          # => false
```
is this one same as:

``` ruby
  validates [1, 2], verify_array: { in: [3, 1] }
```

kgrz replied:
Yes.

iffyuva commented on README.md:
```diff
@@ -125,6 +125,56 @@ class EventForm
+
+
+The only caveat here is that the verification item (symbol/proc) should
+return an Array.
```
in this case, should we have a check in validator verifying the return value of proc?

kgrz replied:
We do. We raise an error if the value of proc is not an Array. Although I'm not sure if we should be raising an error or adding the error to the record.

iffyuva commented on README.md:
```diff
@@ -125,6 +125,56 @@ class EventForm
+validator handles the three cases of `:in`, :not_in` and `:eq`. The
+abstract functionality is as follows:
+
+    validates [1, 2, 3], verify_array: { in: [1, 2, 3, 4] } # => true
```
there is no

``` ruby
  validates :ids, verify_array: true
```

does ^ work?

kgrz replied:
Yes, we have a basic check

iffyuva replied:
i guess, we should update readme in that case.

kgrz replied:
will do that.

iffyuva commented on lib/rails_legit/verify_array_validator.rb:
```diff
@@ -0,0 +1,88 @@
+      record_array_set = SortedSet.new(record_array)
+      verification_array_set = SortedSet.new(verification_array)
+
+      if type_of_comparision == :in
```
i feel there shouldn't be a series of if-elses. is it possible to have combination of comparisons? say, if i want to do something like this:

``` ruby
  validates :ids, verify_array: { in: :user_ids, not_in: :suspended_user_ids }
  #...
  def user_ids
    User.all.map(&:id).map(&:to_s)
  end

  def suspended_user_ids
    # assume suspended is a scope.
    User.suspended.map(&:id).map(&:to_s)
  end
```

kgrz replied:
Hmm, will add some test cases.

kgrz committed:
Add tests for both options provided at once

Earlier, there were no tests for the case where both :in and :not_in
options are specified for a validation. For example,

validates :array,
  verify_array: { in: :one_array, not_in: :another_array }

Added tests covering that case

kgrz committed:
DRY-up spec record objects

kgrz committed:
Update README

- Add explanation about the no-options case for an Array validator
- Clarify :in and :eq validators. These are essentially the same when
  the comparision and given arrays contain the same elements

iffyuva commented:
i tried this one out in one of existing projects. it works, but has one caveat: if i pass a method as a symbol, and if the method is private, validator fails to work.

``` ruby
  validates :ids, verify_array: {in: :user_ids}
  #...

  private
  def user_ids
    User.all.map(&:id).map(&:to_s)
  end
```

can you check?

kgrz committed:
Use a case statement instead of multiple ifelse

kgrz committed:
Fix error message

kgrz committed:
Ensure symbols try to fetch private methods

Earlier, this would fail:

class User < ActiveRecord::Base
  validates :name, verify_array: { in: :valid_names_list }

  private

  def valid_names_list
    ['name1', 'name2']
  end
end

The `private_method_defined?` method can be used to check whether a
private method is defined on the class. So, a method that has a visible
method and a private method can be accessed. If not, we still raise an
error message. This will be modified later to use a much descriptive
message.
